### PR TITLE
Feat/6905 transport grpc

### PR DESCRIPTION
**Pull request by morhag90**
Fixes: amplication/amplication#6905

## PR Details

Add new plugin : transport-grpc.

## PR Checklist

- [ ] Tests for the changes have been added
- [ ] `npm build` doesn't throw any error
- [ ] `npm test` doesn't throw any error

### CODE REVIEW

File: .github/workflows/publish.yaml [hunk 1]
```diff
@@ -27,6 +27,7 @@ on:
           - "observability-opentelemetry"
           - "swagger-apibody"
           - "linter-eslint"
+          - "plugin-transport-grpc"
```
Comment (mulygottlieb) [inline .github/workflows/publish.yaml:30]: Why does this have the "plugin-" prefix unlike all others?

File: plugins/transport-grpc/README.md
Comment (mulygottlieb) [inline plugins/transport-grpc/README.md:1]: This README is completely not relevant, it's copied from the auth-jwt plugin

File: plugins/transport-grpc/package.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "@amplication/plugin-transport-grpc",
+  "version": "0.0.1",
+  "description": "set grpc as protocol for Amplication build",
```
Comment (mulygottlieb) [inline plugins/transport-grpc/package.json:4]: Please change the description to "add gRPC endpoints to your generated services"

File: plugins/transport-grpc/package.json [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  "main": "dist/index.js",
+  "scripts": {
+    "dev": "webpack --watch",
+    "build": "webpack",
+    "prebuild": "rimraf dist",
+    "format:write": "prettier --write \"**/*.{js,ts}\""
+  },
+  "author": "Mor Hagbi",
+  "license": "ISC",
```
Comment (mulygottlieb) [inline plugins/transport-grpc/package.json:13]: I believe all the plugins use an "Apache-2.0" license

File: plugins/transport-grpc/package.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  "devDependencies": {
+    "@amplication/code-gen-types": "2.0.13-beta.1",
```
Comment (mulygottlieb) [inline plugins/transport-grpc/package.json:15]: I think the code-gen-types should move from devDependencies into dependencies. @overbit please opine on this.

File: plugins/transport-grpc/package.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    "typescript": "^5.2.2",
+    "webpack": "^5.88.2",
+    "webpack-cli": "^5.1.4",
+    "webpack-node-externals": "^3.0.0"
+  },
+  "dependencies": {
+    "@amplication/code-gen-utils": "^0.0.6",
+    "pascal-case": "^3.1.2",
+    "prisma-schema-dsl-types": "^1.1.2",
+    "protobuf-dsl": "^0.0.1-beta.20"
```
Comment (mulygottlieb) [inline plugins/transport-grpc/package.json:36]: I think we need to publish a non-beta version of protobuf-dsl once we are happy with what we have there, and then have the dependency here on the non-beta version

File: plugins/transport-grpc/src/core/create-grpc-client-options.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  });
+
+  const packagesArray = packages;
+
+  const protoPathArray = protoPaths;
+
+  const templateMapping = {
+    PACKAGES_ARRAY: packagesArray,
+    PROTO_PATH_ARRAY: protoPathArray,
+    PORT_PATH: builders.stringLiteral("localhost:9090")
```
Comment (mulygottlieb) [inline plugins/transport-grpc/src/core/create-grpc-client-options.ts:50]: Should this be hard-coded here ("localhost:9090") and not arrive from the plugin settings?

File: plugins/transport-grpc/src/constants.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { VariableDictionary } from "@amplication/code-gen-types";
+import { join } from "path";
+
+export const envVariables: VariableDictionary = [
+  { GRPC_CLIENT_URL_PATH: "localhost:9090" },
```
Comment (mulygottlieb) [inline plugins/transport-grpc/src/constants.ts:5]: Is "localhost:9090" later replaced by something arriving from the plugin settings?

File: plugins/transport-grpc/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      relatedEntities
+    );
+
+    await modules.set(protoFile);
+
+    return modules;
+  }
+}
+
+export default JwtAuthPlugin;
```
Comment (mulygottlieb) [inline plugins/transport-grpc/src/index.ts:151]: This does not seem like it belongs here...

File: plugins/transport-grpc/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  createGrpcClientOptionsFile,
+  createGrpcController,
+  createGrpcControllerBase,
+  createGrpcControllerToManyRelationMethods,
+  createGrpcProtoFile,
+} from "./core";
+import { merge } from "lodash";
+import { envVariables } from "./constants";
+
+class JwtAuthPlugin implements AmplicationPlugin {
```
Comment (mulygottlieb) [inline plugins/transport-grpc/src/index.ts:25]: This class name is wrong...

File: plugins/transport-grpc/src/core/create-grpc-client-options.ts (hunk 6 above)
Comment (morhag90) [reply plugins/transport-grpc/src/core/create-grpc-client-options.ts:50]: @mulygottlieb This is the default value for the template mapping.
 In the file, we are using the value from the env file:

image

File: plugins/transport-grpc/src/constants.ts (hunk 7 above)
Comment (morhag90) [reply plugins/transport-grpc/src/constants.ts:5]: @mulygottlieb No, this is the default value configured for the env variable.

**Comment by overbit** [comment]
@mulygottlieb @morhag90 shall we add some unit tests for this? A good example can be this:
https://github.com/amplication/plugins/tree/master/plugins/broker-redis/src/tests

**Review by mulygottlieb — APPROVED** [review]

**Pull request merged**